lib/src/discovery.dart

```diff
@@ -0,0 +1,96 @@
+// Copyright (c) 2015, the Dart project authors.  Please see the AUTHORS file
+// for details. All rights reserved. Use of this source code is governed by a
+// BSD-style license that can be found in the LICENSE file.
+
+import "dart:io";
+
+import "package:path/path.dart" as path;
+
+import "package_config_impl.dart";
+import "package_config_json.dart";
+
+/// Discover the package configuration for a Dart script.
+///
+/// The [baseDirectory] points to the directory of the Dart script.
+/// A package resolution strategy is found by going through the following steps,
+/// and stopping when something is found.
+///
+/// * Check if a `.dart_tool/package_config.json` file exists in the directory.
+/// * Check if a `.packages` file exists in a directory.
+/// * Repeat these checks for the parent directories until reaching the
+///   root directory.
+///
+/// If any of these tests succeed, a `PackageConfig` class is returned.
+/// Returns `null` if no configuration was found. If a configuration
```

Comment:
I definitely depend on the caller to handle errors. This is a tool package that can be embedded in arbitrary other packages, I cannot provide a message that is meaningful for all of them.
As such, returning `null` forces the user to handle the case better than throwing an exception. Using nullable types to represent a "zero or one X" situation is what it's there for.

Reply:
Exceptions can be a lot more specific and can convey a lot more information than `null` (which can't convey anything other than failure).

It is also easier to code around - you can put a whole lot of code inside of a single try/catch block instead of handling every single possible `null` value you might get from every method call you make (if `null` as failure was taken to the extreme).

Reply:
I'm not sure what this extra information would convey. This is just a failure.
The only way to get to returning a `null` is to not find any configuration file.
If a file is found and it's invalid, you'll get format exceptions.